internal/config/config.go

```
@@ -0,0 +1,46 @@
+package config
+
+import (
+	"encoding/json"
+	"errors"
+	"io"
+
+	"github.com/BurntSushi/toml"
+)
+
+type Config interface {
+	FromEnviron() error
+}
+
+const (
+	FromEnviron = iota
+	FromJSON
+	FromTOML
+)
+
+var ErrInvalidWhence = errors.New("invalid whence")
+
+// Load populates config based on whence,
+// if whence is a serialisation format, it is decoded from r.
+func Load(config Config, whence int, r io.Reader) error {
+	switch whence {
+	case FromEnviron:
+		return config.FromEnviron()
+
+	case FromJSON:
+		if r == nil {
+			return errors.New("reader cannot be nil for JSON source")
+		}
```

This guards against a programming error and we should let the default panic behaviour handle this instead as it makes little sense to handle this error at runtime.